pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -1371,20 +1372,30 @@ func (c *Controller) getProxyServiceInstancesByPod(pod *v1.Pod,
 func (c *Controller) GetProxyWorkloadLabels(proxy *model.Proxy) labels.Instance {
 	pod := c.pods.getPodByProxy(proxy)
 	if pod != nil {
-		if _, exist := pod.Labels[model.LocalityLabel]; exist {
-			return pod.Labels
+		var locality, nodeName string
+		locality = c.getPodLocality(pod)
+		if len(proxy.GetNodeName()) == 0 {
+			// this can happen for an "old" proxy which has no `Metadata.NodeName` set
+			// in this case we set the node name in labels on the fly
+			nodeName = pod.Spec.NodeName
```

Comment:
This can be removed in a future release, please add a TODO

Reply:
Added